cloudbridge/providers/openstack/provider.py

```diff
@@ -112,13 +118,26 @@ def _keystone_session(self):
 
         if self._keystone_version == 3:
             from keystoneauth1.identity import v3
-            auth = v3.Password(auth_url=self.auth_url,
-                               username=self.username,
-                               password=self.password,
-                               user_domain_name=self.user_domain_name,
-                               project_domain_id=self.project_domain_id,
-                               project_domain_name=self.project_domain_name,
-                               project_name=self.project_name)
+            if self.username and self.password:
+                auth = v3.Password(auth_url=self.auth_url,
+                                   username=self.username,
+                                   password=self.password,
+                                   user_domain_name=self.user_domain_name,
+                                   project_domain_id=self.project_domain_id,
+                                   project_domain_name=self.project_domain_name,
+                                   project_name=self.project_name)
+            elif self.app_cred_id and self.app_cred_secret:
+                auth = v3.ApplicationCredential(auth_url=self.auth_url,
```

Comment:
This is great. If you get a chance, please update all libs to latest.

Reply:
Done in https://github.com/CloudVE/cloudbridge/pull/297/commits/365284fdba5eabc1157973325ceb39927f909e14